src/typechecker.erl

```diff
@@ -2181,16 +2181,16 @@ do_type_check_expr_in(Env, ResTy, {tuple, _, TS} = Tup) ->
     end;
 
 %% Maps
-do_type_check_expr_in(Env, ResTy, {map, _, Assocs} = Map) ->
+do_type_check_expr_in(Env, ResTy, {map, _, Assocs} = MapCreation) ->
     {AssocTys, VBs, Cs2} = type_check_assocs(Env, Assocs),
-    _MapTy = update_map_type(type(map, any), AssocTys),
-    case subtype(type(map, any), ResTy, Env#env.tenv) of
+    MapTy = update_map_type(type(map, []), AssocTys),
+    case subtype(MapTy, ResTy, Env#env.tenv) of
```

Comment:
Is this correct? I guess you're aware that `{type, _, map, any}` represents `map()` = *any map* and `{type, _, map, []}` represents `#{}` = the empty map type (the type containing only the empty map).

I don't understand exactly what update_map_type does. A description (and possibly a spec) would be useful.

Reply:
I wasn't sure myself what `update_map_type`'s purpose was, but having analysed the map creation and update expressions I now understand it's inferring type of the resulting (created or updated) map, before it can be type-checked.

> I guess you're aware that {type, _, map, any} represents map() = any map and {type, _, map, []} represents #{} = the empty map type (the type containing only the empty map).

Indeed. That's why I renamed `Map` to `MapCreation` and `MapUpdate` accordingly, to make the two clauses more distinct. In case of map creation, we know that the exact type can be fully determined from the analysed expression. Moreover, in case of map creation we know that the inferred type has only non-optional associations (i.e. the fields are already there when we build the map), so the expression's assocs are rewritten (by `update_assocs`).

Reply:
Great. Put some of this as comments in the code. :-)

If the clauses of update_map_type serve different purposes, perhaps it's better to split it into two different functions?

Reply:
> If the clauses of update_map_type serve different purposes, perhaps it's better to split it into two different functions?

I've refactored one of the other clauses somewhat - I think all the clauses do something close enough to each other to keep it a single function.

Reply:
Nice! I'm convinced. Good code comments added. 👍